// react-native-popable/lib/module/Popable.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useCallback, useEffect, useImperativeHandle, useRef, useState } from 'react';
import { Platform // @ts-ignore
, Pressable, StyleSheet, useWindowDimensions, View } from 'react-native';
import Backdrop from './Backdrop';
import Popover from './Popover';
const DEFAULT_LAYOUT = {
  width: 0,
  height: 0,
  x: 0,
  y: 0
};
const Popable = /*#__PURE__*/forwardRef(function Popable({
  action = 'press',
  animated,
  animationType,
  backgroundColor,
  children,
  caret,
  caretPosition,
  content,
  numberOfLines,
  onAction,
  position = 'top',
  strictPosition = false,
  style,
  visible,
  wrapperStyle
}, ref) {
  const dimensions = useWindowDimensions();
  const [popoverVisible, setPopoverVisible] = useState(false);
  const [popoverOffset, setPopoverOffset] = useState({
    left: 0,
    top: 0
  });
  const [popoverLayout, setPopoverLayout] = useState(DEFAULT_LAYOUT);
  const [popoverPagePosition, setPopoverPagePosition] = useState({
    left: 0,
    top: 0
  });
  const [childrenLayout, setChildrenLayout] = useState(DEFAULT_LAYOUT);
  const [computedPosition, setComputedPosition] = useState(position);
  const isInteractive = typeof visible === 'undefined';
  const isPopoverVisible = isInteractive ? popoverVisible : visible;
  const childrenRef = useRef(null);
  const popoverRef = useRef(null);
  useImperativeHandle(ref, () => ({
    show: () => setPopoverVisible(true),
    hide: () => setPopoverVisible(false)
  }));
  const handlers = {};

  if (isInteractive) {
    if (action === 'hover' && Platform.OS === 'web') {
      handlers.onHoverIn = () => {
        setPopoverVisible(true);
        onAction === null || onAction === void 0 ? void 0 : onAction(true);
      };

      handlers.onHoverOut = () => {
        setPopoverVisible(false);
        onAction === null || onAction === void 0 ? void 0 : onAction(false);
      };
    } else if (action === 'press' || action === 'hover' && Platform.OS !== 'web') {
      handlers.onPress = () => {
        if (!visible) {
          var _popoverRef$current;

          (_popoverRef$current = popoverRef.current) === null || _popoverRef$current === void 0 ? void 0 : _popoverRef$current.measure((_x, _y, _width, _height, pageX, pageY) => {
            setPopoverPagePosition({
              left: pageX,
              top: pageY
            });
          });
        }

        onAction === null || onAction === void 0 ? void 0 : onAction(!visible);
        setPopoverVisible(!visible);
      };
    } else {
      handlers.onLongPress = () => {
        onAction === null || onAction === void 0 ? void 0 : onAction(!visible);
        setPopoverVisible(!visible);
      };
    }
  }

  const handleHidePopover = useCallback(() => {
    setPopoverVisible(false);
    onAction === null || onAction === void 0 ? void 0 : onAction(false);
  }, [onAction]);
  const handlePopoverLayout = useCallback(() => {
    var _popoverRef$current2;

    (_popoverRef$current2 = popoverRef.current) === null || _popoverRef$current2 === void 0 ? void 0 : _popoverRef$current2.measureInWindow((x, y, width, height) => {
      setPopoverLayout({
        x,
        y,
        width,
        height
      });
    });
  }, [popoverRef]);
  const handleChildrenLayout = useCallback(() => {
    var _childrenRef$current;

    (_childrenRef$current = childrenRef.current) === null || _childrenRef$current === void 0 ? void 0 : _childrenRef$current.measureInWindow((x, y, width, height) => {
      setChildrenLayout({
        x,
        y,
        width,
        height
      });
    });
  }, [childrenRef]);
  useEffect(() => {
    let nextPosition = position;

    if (!strictPosition) {
      switch (position) {
        case 'left':
          if (popoverLayout.x <= 0) {
            nextPosition = 'right';
          }

          break;

        case 'right':
          if (popoverLayout.x + popoverLayout.width > dimensions.width) {
            nextPosition = 'left';
          }

          break;

        case 'top':
          if (popoverLayout.y <= 0) {
            nextPosition = 'bottom';
          }

          break;

        case 'bottom':
          if (popoverLayout.y + popoverLayout.height >= dimensions.height) {
            nextPosition = 'top';
          }

          break;
      }
    }

    setComputedPosition(nextPosition);
  }, [position, strictPosition, popoverLayout, childrenLayout, dimensions]);
  useEffect(() => {
    let left = 0;
    let top = 0;

    switch (computedPosition) {
      case 'right':
      case 'left':
        top = (popoverLayout.height - childrenLayout.height) / 2;
        break;

      case 'top':
      case 'bottom':
        left = (popoverLayout.width - childrenLayout.width) / 2;
        break;
    }

    setPopoverOffset({
      left,
      top
    });
  }, [computedPosition, popoverLayout, childrenLayout]);
  const sharedPopoverProps = {
    animated,
    animationType,
    backgroundColor,
    caret,
    caretPosition,
    children: content,
    numberOfLines,
    position: computedPosition
  };
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, wrapperStyle]
  }, /*#__PURE__*/React.createElement(Backdrop, {
    visible: isInteractive && popoverVisible,
    onPress: handleHidePopover,
    popoverRef: popoverRef,
    childrenRef: childrenRef
  }, // Backdrop renders the same popover because:
  // since the backdrop adds a layer on top of the screen to
  // detect any "outside popover press", the inner popover becomes
  // unreachable: the upper layer would keep all the touch events.
  // Because the backdrop uses a modal as a layer, we render that
  // same popover inside the modal, and hide the initial one
  // underneath (which explains why the popover below this one has
  // `visible` set to `false`)
  Platform.OS !== 'web' && /*#__PURE__*/React.createElement(Popover, _extends({}, sharedPopoverProps, {
    forceInitialAnimation: true,
    visible: isPopoverVisible,
    style: [{
      position: 'absolute',
      transform: [{
        translateX: popoverPagePosition.left
      }, {
        translateY: popoverPagePosition.top
      }]
    }, style]
  }))), /*#__PURE__*/React.createElement(Popover, _extends({
    ref: popoverRef
  }, sharedPopoverProps, {
    onLayout: handlePopoverLayout,
    visible: Platform.OS === 'web' ? isPopoverVisible : false,
    style: [computedPosition === 'top' && styles.popoverTop, computedPosition === 'bottom' && styles.popoverBottom, computedPosition === 'left' && {
      alignItems: 'flex-end',
      right: childrenLayout.width
    }, computedPosition === 'right' && {
      left: childrenLayout.width
    }, {
      position: 'absolute',
      transform: [{
        translateX: popoverOffset.left * -1
      }, {
        translateY: popoverOffset.top * -1
      }]
    }, style]
  })), /*#__PURE__*/React.createElement(Pressable, _extends({
    ref: childrenRef,
    onLayout: handleChildrenLayout
  }, handlers), children));
});
const styles = StyleSheet.create({
  container: {
    position: 'relative',
    zIndex: 1
  },
  popoverTop: {
    bottom: '100%'
  },
  popoverBottom: {
    top: '100%'
  }
});
export default Popable;
