// react-native-tuvali/lib/commonjs/types/events.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VerificationStatus = exports.EventTypes = void 0;
/* This file redefines the types from Native Modules to allow auto complete to the consumer applications,
  However this means, if the native module type definition changes, such changes needs to be applied here as well
  to avoid mismatch of type information
*/
let EventTypes;
exports.EventTypes = EventTypes;
(function (EventTypes) {
  EventTypes["onConnected"] = "onConnected";
  EventTypes["onSecureChannelEstablished"] = "onSecureChannelEstablished";
  EventTypes["onDataSent"] = "onDataSent";
  EventTypes["onDataReceived"] = "onDataReceived";
  EventTypes["onVerificationStatusReceived"] = "onVerificationStatusReceived";
  EventTypes["onError"] = "onError";
  EventTypes["onDisconnected"] = "onDisconnected";
})(EventTypes || (exports.EventTypes = EventTypes = {}));
let VerificationStatus;
exports.VerificationStatus = VerificationStatus;
(function (VerificationStatus) {
  VerificationStatus[VerificationStatus["ACCEPTED"] = 0] = "ACCEPTED";
  VerificationStatus[VerificationStatus["REJECTED"] = 1] = "REJECTED";
})(VerificationStatus || (exports.VerificationStatus = VerificationStatus = {}));
