// react-native-popable/lib/commonjs/Caret.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _styleGuide = require("./style-guide");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = ({
  align,
  backgroundColor,
  position,
  style
}) => {
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, align === 'center' && styles.containerCenter, align === 'right' && styles.containerRight, !!backgroundColor && {
      backgroundColor
    }, position === 'top' && styles.containerPositionTop, position === 'bottom' && styles.containerPositionBottom, position === 'left' && styles.containerPositionLeft, position === 'right' && styles.containerPositionRight, style]
  });
};

exports.default = _default;

const styles = _reactNative.StyleSheet.create({
  container: {
    width: _styleGuide.CARET_SIDE_SIZE,
    height: _styleGuide.CARET_SIDE_SIZE,
    backgroundColor: _styleGuide.POPOVER_BACKGROUND_COLOR,
    transform: [{
      rotate: '45deg'
    }],
    borderRadius: _styleGuide.BORDER_RADIUS
  },
  containerPositionTop: {
    marginTop: (_styleGuide.CARET_SIDE_SIZE / 2 + _styleGuide.BORDER_RADIUS / 2) * -1,
    marginBottom: _styleGuide.CARET_SIDE_SIZE / 2 + _styleGuide.BORDER_RADIUS / 2
  },
  containerPositionBottom: {
    marginBottom: (_styleGuide.CARET_SIDE_SIZE / 2 + _styleGuide.BORDER_RADIUS / 2) * -1,
    marginTop: _styleGuide.CARET_SIDE_SIZE / 2 + _styleGuide.BORDER_RADIUS / 2
  },
  containerPositionLeft: {
    marginLeft: (_styleGuide.CARET_SIDE_SIZE / 2 + _styleGuide.BORDER_RADIUS / 2) * -1,
    marginRight: _styleGuide.CARET_SIDE_SIZE / 2 + _styleGuide.BORDER_RADIUS / 2
  },
  containerPositionRight: {
    marginRight: (_styleGuide.CARET_SIDE_SIZE / 2 + _styleGuide.BORDER_RADIUS / 2) * -1,
    marginLeft: _styleGuide.CARET_SIDE_SIZE / 2 + _styleGuide.BORDER_RADIUS / 2
  },
  containerCenter: {
    alignSelf: 'center'
  },
  containerRight: {
    alignSelf: 'flex-end'
  }
});
